lib/core/utils/preload-cssom.js

```diff
@@ -0,0 +1,205 @@
+/**
+ * Returns a then(able) queue of CSSStyleSheet(s)
+ * @param {Object} ownerDocument document object to be inspected for stylesheets
+ * @param {number} timeout on network request for stylesheet that need to be externally fetched
+ * @param {Function} convertTextToStylesheetFn a utility function to generate a style sheet from text
+ * @return {Object} queue
+ * @private
+ */
+function loadCssom({ root, shadowId }, timeout, convertTextToStylesheetFn) {
+	/**
+	 * Make an axios get request to fetch a given resource and resolve
+	 * @method getExternalStylesheet
+	 * @private
+	 * @param {Object} param an object with properties to configure the external XHR
+	 * @property {Object} param.resolve resolve callback on queue
+	 * @property {Object} param.reject reject callback on queue
+	 * @property {String} param.url string representing the url of the resource to load
+	 * @property {Number} param.timeout timeout to about network call
+	 */
+	function getExternalStylesheet({ resolve, reject, url }) {
+		axe.imports
+			.axios({
+				method: 'get',
+				url,
+				timeout
+			})
+			.then(({ data }) => {
+				const sheet = convertTextToStylesheetFn({
+					data,
+					isExternal: true,
+					shadowId
+				});
+				resolve(sheet);
+			})
+			.catch(reject);
+	}
+
+	const q = axe.utils.queue();
+
+	// iterate to decipher multi-level nested sheets if any (this is essential to retrieve styles from shadowDOM)
+	Array.from(root.styleSheets).forEach(sheet => {
+		// ignore disabled sheets
+		if (sheet.disabled) {
+			return;
+		}
+		// attempt to retrieve cssRules, or for external sheets make a XMLHttpRequest
+		try {
+			// accessing .cssRules throws for external (cross-domain) sheets, which is handled in the catch
+			const cssRules = sheet.cssRules;
+			// read all css rules in the sheet
+			const rules = Array.from(cssRules);
+
+			// filter rules that are included by way of @import or nested link
+			const importRules = rules.filter(r => r.href);
+
+			// if no import or nested link rules, with in these cssRules
+			// return current sheet
+			if (!importRules.length) {
+				q.defer(resolve =>
+					resolve({
+						sheet,
+						isExternal: false,
+						shadowId
+					})
+				);
+				return;
+			}
+
+			// if any import rules exists, fetch via `href` which eventually constructs a sheet with results from resource
+			importRules.forEach(rule => {
+				q.defer((resolve, reject) => {
+					getExternalStylesheet({ resolve, reject, url: rule.href });
+				});
+			});
+
+			// in the same sheet -  get inline rules in <style> tag or in a CSSStyleSheet excluding @import or nested link
+			const inlineRules = rules.filter(rule => !rule.href);
+
+			// concat all cssText into a string for inline rules
+			const inlineRulesCssText = inlineRules
+				.reduce((out, rule) => {
```

Comment:
This `.reduce()` seems to be performing the function of `.map()`. Should we refactor to use the intended native method here?

Reply:
I am taking each `CSSRule` and appending its `cssText` in to an array - `out`. I intend to keep this as `reduce`, as is, because I believe there will be logic going into this iteration as I tackle for `@import` and `media` rules.

Appreciate the comment.